sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala

```diff
@@ -0,0 +1,515 @@
+          };
+
+          record @serializable S = {};
+
+          template (this : S) = {
+            precondition True;
+            signatories Nil @Party;
+            observers Nil @Party;
+            choice NoOpS (self) (arg: Mod:NoOpT) : Unit,
+                controllers Mod:NoOpT {controllers} arg,
+                observers Nil @Party
+              to upure @Unit ();
+          };
+
+          record @serializable T = { signatories: List Party, cid: Option (ContractId Mod:T) };
+          record @serializable NoOpT = { controllers: List Party, cid: Option (ContractId Mod:T) };
+          record @serializable Key = { signatories: List Party, cid: Option (ContractId Mod:T) };
+
+          template (this : T) = {
+            precondition True;
+            signatories Mod:T {signatories} this;
+            observers Nil @Party;
+            choice NoOpT (self) (arg: Mod:NoOpT) : Unit,
+                controllers Mod:NoOpT {controllers} arg,
+                observers Nil @Party
+              to upure @Unit ();
+            implements Mod:I {
+              view = Mod:EmptyView {};
+            };
+            key @Mod:Key
+              (Mod:Key { signatories = Mod:T {signatories} this, cid = None @(ContractId Mod:T) })
+              (\ (key: Mod:Key) -> Mod:Key {signatories} key);
+          };
+
+        }
+    """
+
+  private[this] val compiledPackage = ConcurrentCompiledPackages(
+    Compiler.Config.Default(majorLanguageVersion)
+  )
+  assert(compiledPackage.addPackage(pkgId, pkg) == ResultDone.Unit)
+
+  val parties = collection.immutable.TreeSet[Ref.Party]("Alice")
+  private[this] val valueParties = ValueList(parties.iterator.map(ValueParty).to(FrontStack))
+
+  "preprocessCommand" should {
+
+    val suffixedCid = ContractId.V1.assertBuild(
+      crypto.Hash.hashPrivateKey("a suffixed V1 Contract ID"),
+      Bytes.assertFromString("00"),
+    )
+
+    val nonSuffixedCid = ContractId.V1.assertBuild(
+      crypto.Hash.hashPrivateKey("a non-suffixed V1 Contract ID"),
+      Bytes.Empty,
+    )
+
+    val someSuffixedCid = ValueOptional(Some(ValueContractId(suffixedCid)))
+    val someNonSuffixedCid = ValueOptional(Some(ValueContractId(nonSuffixedCid)))
+
+    val TId: Ref.TypeConName = "Mod:T"
+    val TName: Ref.TypeConRef = Ref.TypeConRef(Ref.PackageRef.Name(pkg.pkgName), TId.qualifiedName)
+
+    val IId: Ref.TypeConName = "Mod:I"
+    val IName: Ref.TypeConRef = Ref.TypeConRef(Ref.PackageRef.Name(pkg.pkgName), IId.qualifiedName)
+
+    val normalizedRecordSome = ValueRecord("", ImmArray("" -> valueParties, "" -> someSuffixedCid))
+    val normalizedRecordNone = ValueRecord("", ImmArray("" -> valueParties))
+    val nonNormalizedRecordNone = ValueRecord("", ImmArray("" -> valueParties, "" -> ValueNone))
+    val nonNormalizedRecordUpgrade =
+      ValueRecord("", ImmArray("" -> valueParties, "" -> ValueNone, "" -> ValueNone))
+    val illTypedRecord = ValueRecord("", ImmArray("" -> valueParties, "" -> ValueUnit))
```

Comment:
As a special case of illTypedRecord we could also have a record that does not have enough non-optional fields.

Reply:
This should be capture by the tests for valueTranslator.